wgpu-core/src/resource.rs

```diff
@@ -160,12 +160,47 @@ pub struct Buffer<B: hal::Backend> {
     pub(crate) device_id: Stored<DeviceId>,
     pub(crate) usage: wgt::BufferUsage,
     pub(crate) size: wgt::BufferAddress,
-    pub(crate) full_range: (),
+    // An allocated range in this allocator means that the range in question is NOT yet initialized.
+    pub(crate) uninitialized_ranges: range_alloc::RangeAllocator<wgt::BufferAddress>,
     pub(crate) sync_mapped_writes: Option<hal::memory::Segment>,
     pub(crate) life_guard: LifeGuard,
     pub(crate) map_state: BufferMapState<B>,
 }
 
+impl<B: hal::Backend> Buffer<B> {
+    pub(crate) fn uninitialized_ranges_in_range<
+        R: std::iter::FromIterator<Range<wgt::BufferAddress>>,
```

Comment:
could return `impl Iterator<...>` instead of this generic magic (minus notes that I wrote in an earlier comment)